python/cudf/cudf/core/column/categorical.py

```diff
@@ -1068,51 +1068,33 @@ def notnull(self) -> ColumnBase:
 
         return result
 
-    def fillna(
-        self,
-        fill_value: Any = None,
-        method: str | None = None,
-    ) -> Self:
-        """
-        Fill null values with *fill_value*
-        """
-        if fill_value is not None:
-            fill_is_scalar = np.isscalar(fill_value)
-
-            if fill_is_scalar:
-                if fill_value == _DEFAULT_CATEGORICAL_VALUE:
-                    fill_value = self.codes.dtype.type(fill_value)
-                else:
-                    try:
-                        fill_value = self._encode(fill_value)
-                        fill_value = self.codes.dtype.type(fill_value)
-                    except ValueError as err:
-                        err_msg = "fill value must be in categories"
-                        raise ValueError(err_msg) from err
-            else:
-                fill_value = column.as_column(fill_value, nan_as_null=False)
-                if isinstance(fill_value, CategoricalColumn):
-                    if self.dtype != fill_value.dtype:
-                        raise TypeError(
-                            "Cannot set a Categorical with another, "
-                            "without identical categories"
-                        )
-                # TODO: only required if fill_value has a subset of the
-                # categories:
-                fill_value = fill_value._set_categories(
-                    self.categories,
-                    is_unique=True,
-                )
-                fill_value = column.as_column(fill_value.codes).astype(
-                    self.codes.dtype
+    def _validate_fillna_value(
+        self, fill_value: ScalarLike | ColumnLike
+    ) -> cudf.Scalar | ColumnBase:
+        """Align fill_value for .fillna based on column type."""
+        if cudf.api.types.is_scalar(fill_value):
+            if fill_value != _DEFAULT_CATEGORICAL_VALUE:
+                try:
+                    fill_value = self._encode(fill_value)
+                except ValueError as err:
+                    raise ValueError(
+                        f"{fill_value=} must be in categories"
+                    ) from err
+            return cudf.Scalar(fill_value, dtype=self.codes.dtype)
+        else:
+            fill_value = column.as_column(fill_value, nan_as_null=False)
+            if (
+                isinstance(fill_value.dtype, CategoricalDtype)
+                and self.dtype != fill_value.dtype
+            ):
+                raise TypeError(
+                    "Cannot set a Categorical with another, "
+                    "without identical categories"
                 )
-
-        # Validation of `fill_value` will have to be performed
-        # before returning self.
-        if not self.nullable:
-            return self
-
-        return super().fillna(fill_value, method=method)
+            fill_value = fill_value._set_categories(
+                self.categories,
+            )
```

Comment:
*issue*: I am still worried about this. Whose job is it to confirm that after the `as_column` call, `fill_value` is a `CategoricalColumn` (which is the only object for which the `_set_categories` method is defined).

I can see this sort of mimics the logic of before, but I don't think that logic made any sense either.

AFAICT, there are four cases:

1. fill_value is scalar no problem, goes down the other branch
2. fill_value is CategoricalColumn with matching dtype
3. fill_value is CategoricalColumn with mismatching dtype
4. fill_value is Column, but not categorical

Case 2 doesn't satisfy `is_categorical and fill_value.dtype != self.dtype` (so no `TypeError`), but then, by construction, `self.categories` must _equal_ `fill_value.categories`, and so the `_set_categories` call is surely unnecessary.

Case 3 _does_ satisfy `is_categorical and fill_value.dtype != self.dtype`, and therefore raises `TypeError`

Case 4 _doesn't_ satisfy `is_categorical` (so no `TypeError`) but then we unconditionally go forward and try and call `_set_categories` which raises `AttributeError`.

IOw, I think we should get a nicer error message for the case:

```
s = cudf.Series([1, 2, None, None]).astype("category")

s.fillna(cudf.Series([None, None, 1, 2]))
```
(This currently raises an `AttributeError` on an internal missing method).

So I think the logic wants to be:

```
fill_value = column.as_column(fill_value, nan_as_null=False)

if isinstance(fill_value.dtype, CategoricalDtype):
    if self.dtype != fill_value.dtype:
        raise TypeError("Cannot set a categorical with another without identical categories")
else:
    raise TypeError("Cannot set a categorical with non-categorical data")
```

Or, if we need to support arbitrary array-likes, then we should cast `fill_value` to the correct dtype. And we probably need to then validate that this works fine, or something.

Reply:
Yeah that's a good point, might as well modify this input validation while I'm refactoring fillna. I changed this block to raise `TypeError`s like you suggseted

Reply:
Thanks!